src/vos/vos_dtx.c

```diff
@@ -852,18 +918,39 @@ vos_dtx_alloc(struct umem_instance *umm, struct dtx_handle *dth,
 	dtx->te_sec = crt_hlc_get();
 	dtx->te_records = rec_umoff;
 	dtx->te_next = UMOFF_NULL;
-	dtx->te_prev = UMOFF_NULL;
 
 	rbund.trb_umoff = dtx_umoff;
 	d_iov_set(&riov, &rbund, sizeof(rbund));
 	d_iov_set(&kiov, &dth->dth_xid, sizeof(dth->dth_xid));
 	rc = dbtree_upsert(cont->vc_dtx_active_hdl, BTR_PROBE_EQ,
 			   DAOS_INTENT_UPDATE, &kiov, &riov);
-	if (rc == 0) {
-		dth->dth_ent = dtx_umoff;
-		*dtxp = dtx;
-	}
+	if (rc)
+		goto out;
+
+	dth->dth_ent = dtx_umoff;
+	*dtxp = dtx;
+
+	tab = &cont->vc_cont_df->cd_dtx_table_df;
+	umem_tx_add_ptr(umm, tab, sizeof(*tab));
+
+	dtx->te_prev = tab->tt_active_tail;
+	if (dtx_is_null(tab->tt_active_tail)) {
+		D_ASSERTF(cont->vc_cont_df->cd_lpe == DAOS_EPOCH_MAX,
+			  DF_U64" "DF_U64,
+			  cont->vc_cont_df->cd_lpe, dtx->te_epoch);
+		cont->vc_cont_df->cd_lpe = dtx->te_epoch;
+	} else {
+		D_ASSERTF(cont->vc_cont_df->cd_lpe <= dtx->te_epoch,
+			  DF_U64" "DF_U64,
+			  cont->vc_cont_df->cd_lpe, dtx->te_epoch);
 
+		ent = umem_off2ptr(umm, tab->tt_active_tail);
+		umem_tx_add_ptr(umm, &ent->te_next, sizeof(ent->te_next));
+
+		ent->te_next = dtx_umoff;
+	}
+	tab->tt_active_tail = dtx_umoff;
```

Comment:
There may be CPU yield between generating the epoch and allocating the DTX entry. So older epoch does NOT means related DTX entry will be allocated earlier than the one with newer epoch.